File: 0547-number-of-provinces/0547-number-of-provinces.java
public class Solution {
    public void DFS(Map<Integer, List<Integer>> adj, int i, boolean[] visited) {
        visited[i] = true;
        for (int v : adj.get(i)) {
            if (!visited[v]) {
                DFS(adj, v, visited);
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        int count = 0;
        int n = isConnected.length;
        boolean[] visited = new boolean[n];
        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int i = 0; i < n; i++) {
            adj.put(i, new ArrayList<>());
            for (int j = 0; j < n; j++) {
                if (isConnected[i][j] == 1) {
                    adj.get(i).add(j);
                    adj.get(j).add(i);
                }
            }
        }
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                DFS(adj, i, visited);
                count++;
            }
        }
        return count;
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def DFS(self, adj, i, visited):
        visited[i] = True
        for v in adj[i]:
            if not visited[v]:
                self.DFS(adj, v, visited)

    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        count = 0
        n = len(isConnected)
        visited = [False] * n
        adj = {}
        for i in range(n):
            adj[i] = []
            for j in range(n):
                if isConnected[i][j] == 1:
                    if j not in adj:
                        adj[j] = []
                    adj[i].append(j)
                    adj[j].append(i)
        for i in range(n):
            if not visited[i]:
                self.DFS(adj, i, visited)
                count += 1
        return count